lib/vector-core/src/event/metric/mod.rs

```diff
@@ -0,0 +1,1008 @@
+        self.data.value.fmt(fmt)
+    }
+}
+
+impl EventDataEq for Metric {
+    fn event_data_eq(&self, other: &Self) -> bool {
+        self.series == other.series
+            && self.data == other.data
+            && self.metadata.event_data_eq(&other.metadata)
+    }
+}
+
+impl ByteSizeOf for Metric {
+    fn allocated_bytes(&self) -> usize {
+        self.series.allocated_bytes()
+            + self.data.allocated_bytes()
+            + self.metadata.allocated_bytes()
+    }
+}
+
+impl Finalizable for Metric {
+    fn take_finalizers(&mut self) -> EventFinalizers {
+        self.metadata.take_finalizers()
+    }
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, PartialOrd, Serialize)]
+#[serde(rename_all = "snake_case")]
+/// A metric may be an incremental value, updating the previous value of
+/// the metric, or absolute, which sets the reference for future
+/// increments.
+pub enum MetricKind {
+    Incremental,
+    Absolute,
+}
+
+#[cfg(feature = "vrl")]
+impl TryFrom<vrl_lib::Value> for MetricKind {
+    type Error = String;
+
+    fn try_from(value: vrl_lib::Value) -> Result<Self, Self::Error> {
+        let value = value.try_bytes().map_err(|e| e.to_string())?;
+        match std::str::from_utf8(&value).map_err(|e| e.to_string())? {
+            "incremental" => Ok(Self::Incremental),
+            "absolute" => Ok(Self::Absolute),
+            value => Err(format!(
+                "invalid metric kind {}, metric kind must be `absolute` or `incremental`",
+                value
+            )),
+        }
+    }
+}
+
+#[cfg(feature = "vrl")]
+impl From<MetricKind> for vrl_lib::Value {
+    fn from(kind: MetricKind) -> Self {
+        match kind {
+            MetricKind::Incremental => "incremental".into(),
+            MetricKind::Absolute => "absolute".into(),
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! samples {
+    ( $( $value:expr => $rate:expr ),* ) => {
+        vec![ $( crate::event::metric::Sample { value: $value, rate: $rate }, )* ]
+    }
+}
+
+#[macro_export]
+macro_rules! buckets {
+    ( $( $limit:expr => $count:expr ),* ) => {
+        vec![ $( crate::event::metric::Bucket { upper_limit: $limit, count: $count }, )* ]
+    }
+}
+
+#[macro_export]
+macro_rules! quantiles {
+    ( $( $q:expr => $value:expr ),* ) => {
+        vec![ $( crate::event::metric::Quantile { quantile: $q, value: $value }, )* ]
+    }
+}
+
+pub fn zip_samples(
+    values: impl IntoIterator<Item = f64>,
+    rates: impl IntoIterator<Item = u32>,
+) -> Vec<Sample> {
+    values
+        .into_iter()
+        .zip(rates.into_iter())
+        .map(|(value, rate)| Sample { value, rate })
+        .collect()
+}
+
+pub fn zip_buckets(
+    limits: impl IntoIterator<Item = f64>,
+    counts: impl IntoIterator<Item = u32>,
+) -> Vec<Bucket> {
+    limits
+        .into_iter()
+        .zip(counts.into_iter())
+        .map(|(upper_limit, count)| Bucket { upper_limit, count })
+        .collect()
+}
+
+pub fn zip_quantiles(
```

Comment:
I think this could be `pub(crate)` and potentially a good candidate for being marked for inlining. Same for the other zip functions here.

Reply:
Good call. Addressed via 0945e10.